src/DotNetty.Buffers/ByteBufferUtil.cs

```diff
@@ -570,6 +570,56 @@ static void AppendHexDumpRowPrefix(StringBuilder dump, int row, int rowStartInde
             }
         }
 
+
+        /// <summary>
+        /// Encode the given {@link CharBuffer} using the given {@link Charset} into a new {@link ByteBuf} which
+        /// is allocated via the {@link ByteBufAllocator}.
+        /// </summary>
+        /// <param name="alloc">The {@link IByteBufferAllocator} to allocate {@link IByteBuffer}.</param>
+        /// <param name="src">src The {@link String} to encode.</param>
+        /// <param name="encoding">charset The specified {@link Encoding}</param>
+        public static IByteBuffer EncodeString(IByteBufferAllocator alloc, string src, Encoding encoding)
+        {
+            return EncodeString0(alloc, src, encoding, 0);
+        }
+
+        /// <summary>
+        /// Encode the given {@link CharBuffer} using the given {@link Charset} into a new {@link ByteBuf} which
+        /// is allocated via the {@link ByteBufAllocator}.
+        /// </summary>
+        /// <param name="alloc">The {@link IByteBufferAllocator} to allocate {@link IByteBuffer}.</param>
+        /// <param name="src">src The {@link String} to encode.</param>
+        /// <param name="encoding">charset The specified {@link Encoding}</param>
+        /// <param name="extraCapacity">the extra capacity to alloc except the space for decoding.</param>
+        public static IByteBuffer EncodeString(IByteBufferAllocator alloc, string src, Encoding encoding, int extraCapacity)
+        {
+            return EncodeString0(alloc, src, encoding, extraCapacity);
+        }
+
+        static IByteBuffer EncodeString0(IByteBufferAllocator alloc, string src, Encoding encoding, int extraCapacity)
+        {
+            var bytes = encoding.GetBytes(src);
```

Comment:
until we've introduced analog of `internalNioBuffer` (which we'll likely have), it should be safe to assume byte buffer from allocator has `HasArray == true` meaning it is safe to operate of its backing array directly (through `buf.Array`). then this piece can be put together with less memory ops:
```
        int length = encoding.GetMaxByteCount(src.Length) + extraCapacity;
        ...
        var dst = alloc.Buffer(length);
        Contract.Assert(dst.HasArray, "Operation expects allocator to operate array-based buffers.");
        encoding.GetBytes(src, 0, src.Length, dst.Array, dst.ArrayOffset);
        ...
```